Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,222 @@
+      let messageDeframer = GRPCMessageDeframer(
+        maximumPayloadSize: maximumPayloadSize,
+        decompressor: decompressor
+      )
+      self.deframer = NIOSingleStepByteToMessageProcessor(messageDeframer)
+      self.compressor = compressor
+    }
+  }
+  
+  init(configuration: Configuration) {
+    self.state = .clientIdleServerIdle(configuration)
+  }
+  
+  mutating func send(metadata: Metadata) throws {
+    // Client only sends metadata when opening.
+    try self.clientSentMetadata()
+  }
+  
+  mutating func send(message: [UInt8]) {
+    // Client sends message.
+    switch self.state {
+    case .clientIdleServerIdle(let configuration):
+      preconditionFailure("Client not yet open")
+    case .clientOpenServerIdle(var clientOpenState):
+      clientOpenState.framer.append(message)
+      self.state = .clientOpenServerClosed(clientOpenState)
+    case .clientOpenServerOpen(var clientOpenState):
+      clientOpenState.framer.append(message)
+      self.state = .clientOpenServerClosed(clientOpenState)
+    case .clientOpenServerClosed:
+      // Nothing to do: no point in sending a message if the server's closed.
+      ()
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      preconditionFailure("Client is closed, cannot send a message")
+    }
+  }
+  
+  mutating func send(status: String, trailingMetadata: Metadata) throws {
+    // Only server does this.
+  }
+  
+  /// Returns the client's next request to the server.
+  /// - Returns: The request to be made to the server.
+  mutating func nextRequest() throws -> ByteBuffer? {
+    switch self.state {
+    case .clientIdleServerIdle(let configuration):
+      preconditionFailure("Client is not open yet.")
+    case .clientOpenServerIdle(var clientOpenState):
+      let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+      self.state = .clientOpenServerIdle(clientOpenState)
+      return request
+    case .clientOpenServerOpen(var clientOpenState):
+      let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+      self.state = .clientOpenServerOpen(clientOpenState)
+      return request
+    case .clientOpenServerClosed:
+      // Nothing to do: no point in sending request if server is closed.
+      return nil
+    case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+      preconditionFailure("Can't send request if client is closed")
+    }
+  }
+  
+  mutating func receive(metadata: Metadata, endStream: Bool) {
+    // This is metadata received by the client from the server.
+    // It can be initial, which confirms that the server is now open;
+    // or an END_STREAM trailer, meaning the response is over.
+    if endStream {
+      self.clientReceivedEndHeader()
+    } else {
+      self.clientReceivedMetadata()
+    }
+  }
+  
+  mutating func receive(message: ByteBuffer, messageProcessor: ([UInt8]) throws -> Void) throws {
```

Comment:
Hmm not sure about the message processor, I think instead our processor should buffer messages then when we get channel read complete we can read them out in a similar way to how we handle messages to send.

Reply:
Ah yeah forgot to leave a comment here. I definitely don't like this, just added it temporarily until I came up with something better. It's the downside of using the NIO processor, we need to fit things to that API. Buffering at this level feels like the only thing we can do here really, so I'll go for that.